phoenix-core/src/main/java/org/apache/phoenix/util/CursorUtil.java

```diff
@@ -0,0 +1,203 @@
+ */
+
+package org.apache.phoenix.util;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.phoenix.compile.QueryPlan;
+import org.apache.phoenix.compile.OrderByCompiler.OrderBy;
+import org.apache.phoenix.execute.CursorFetchPlan;
+import org.apache.phoenix.iterate.CursorResultIterator;
+import org.apache.phoenix.parse.CloseStatement;
+import org.apache.phoenix.parse.DeclareCursorStatement;
+import org.apache.phoenix.parse.OpenStatement;
+import org.apache.phoenix.schema.tuple.Tuple;
+
+public final class CursorUtil {
+
+    private static class CursorWrapper {
+        private final String cursorName;
+        private final String selectSQL;
+        private boolean isOpen = false;
+        QueryPlan queryPlan;
+        ImmutableBytesWritable row;
+        ImmutableBytesWritable previousRow;
+        private Scan scan;
+        private boolean moreValues=true;
+        private boolean isReversed;
+        private boolean islastCallNext;
+        private CursorFetchPlan fetchPlan;
+        private int offset = -1;
+
+        private CursorWrapper(String cursorName, String selectSQL, QueryPlan queryPlan){
+            this.cursorName = cursorName;
+            this.selectSQL = selectSQL;
+            this.queryPlan = queryPlan;
+            this.islastCallNext = true;
+            this.fetchPlan = new CursorFetchPlan(queryPlan);
+        }
+
+        private synchronized void openCursor(Connection conn) throws SQLException {
+            if(isOpen){
+                return;
+            }
+            this.scan = this.queryPlan.getContext().getScan();
+            isReversed=OrderBy.REV_ROW_KEY_ORDER_BY.equals(this.queryPlan.getOrderBy());
+            isOpen = true;
+        }
+
+        private void closeCursor() throws SQLException {
+            isOpen = false;
+            ((CursorResultIterator) fetchPlan.iterator()).closeCursor();
+            //TODO: Determine if the cursor should be removed from the HashMap at this point.
+            //Semantically it makes sense that something which is 'Closed' one should be able to 'Open' again.
+            mapCursorIDQuery.remove(this.cursorName);
+        }
+
+        private QueryPlan getFetchPlan(boolean isNext, int fetchSize) throws SQLException {
+            if (!isOpen)
+                throw new SQLException("Fetch call on closed cursor '" + this.cursorName + "'!");
+            ((CursorResultIterator)fetchPlan.iterator()).setFetchSize(fetchSize);
+            if (!queryPlan.getStatement().isAggregate() || !queryPlan.getStatement().isDistinct()) { 
+            	if (islastCallNext != isNext) {
+                    if (islastCallNext && !isReversed){
+                       ScanUtil.setReversed(scan);
+                } else {
+                       ScanUtil.unsetReversed(scan);
+                }
```

Comment:
this code seems to be for reverse/prior and belongs to another JIRA. can we remove this if it can affect the functionality?